### DISCUSSION TITLE
ASM1074 USB3 HUB power control

### DISCUSSION THREAD

**Discussion by tikiss** [mvp/uhubctl#564] [Q&A] [answered]
I am using an ASM1074 USB3 4 port HUB on a custom board with a Compute Module 4 from Raspberry. The HUB is working fine, but I want to switch off and on the USB3 ports power (+5V), `lsusb -v` give this for the USB3 part of the HUB:
```
  wHubCharacteristic 0x0009
    Per-port power switching
    Per-port overcurrent protection
```
and gives this for the USB2 part of the HUB:
```
  wHubCharacteristic 0x00a9
    Per-port power switching
    Per-port overcurrent protection
    TT think time 16 FS bits
    Port indicators
```
I have found your tool to control the USB3 HUB port power, but unfortunately, I cannot achieve what I want.
`uhubctl` gives the following back:
```
Current status for hub 3-1 [0424:2512, USB 2.00, 2 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
Current status for hub 3 [1d6b:0002 Linux 6.1.21-v8+ dwc2_hsotg DWC OTG Controller fe980000.usb, USB 2.00, 1 ports, ppps]
  Port 1: 0507 power highspeed suspend enable connect [0424:2512, USB 2.00, 2 ports, ppps]
Current status for hub 2-2 [174c:3074 Asmedia ASM107x 0000000A0001, USB 3.00, 4 ports, ppps]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 0203 power 5gbps U0 enable connect [0781:5583 SanDisk Ultra Fit 01019f68043bf5c19cf7e14ab43fb36e78cddaa92af2b1c8fc0be3e508fcf9a]
Current status for hub 1-2 [174c:2074 Asmedia ASM107x 0000000A0001, USB 2.10, 4 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0100 power
```
On port 4 I have connected a mass storage device, so I sent the following command to power it off: 
`sudo uhubctl -l 2-2 -a off -p 4`
And got back this after `uhubctl`:
```
Current status for hub 3-1 [0424:2512, USB 2.00, 2 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
Current status for hub 3 [1d6b:0002 Linux 6.1.21-v8+ dwc2_hsotg DWC OTG Controller fe980000.usb, USB 2.00, 1 ports, ppps]
  Port 1: 0507 power highspeed suspend enable connect [0424:2512, USB 2.00, 2 ports, ppps]
Current status for hub 2-2 [174c:3074 Asmedia ASM107x 0000000A0001, USB 3.00, 4 ports, ppps]
  Port 1: 02a0 power 5gbps Rx.Detect
  Port 2: 02a0 power 5gbps Rx.Detect
  Port 3: 02a0 power 5gbps Rx.Detect
  Port 4: 00a0 off
Current status for hub 1-2 [174c:2074 Asmedia ASM107x 0000000A0001, USB 2.10, 4 ports, ppps]
  Port 1: 0100 power
  Port 2: 0100 power
  Port 3: 0100 power
  Port 4: 0000 off
```
So it seams works, if I perform: `lsusb -t`:
```
/:  Bus 03.Port 1: Dev 1, Class=root_hub, Driver=dwc2/1p, 480M
    |__ Port 1: Dev 2, If 0, Class=Hub, Driver=hub/2p, 480M
/:  Bus 02.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 10000M
    |__ Port 1: Dev 3, If 0, Class=Mass Storage, Driver=usb-storage, 5000M
    |__ Port 2: Dev 2, If 0, Class=Hub, Driver=hub/4p, 5000M
/:  Bus 01.Port 1: Dev 1, Class=root_hub, Driver=xhci_hcd/2p, 480M
    |__ Port 2: Dev 3, If 0, Class=Hub, Driver=hub/4p, 480M
```
The mass storage is disappeared, but if I measure the port voltage, I can measure 5V. So I do not understand, how can I switch of the port 5V?
Just some more information, every port of the HUB has an over-current protection IC, which is connected to the USB3 HUB, so if I make a short circuit on one port of the USB3 HUB (e.g.: on port 4), the port power is turned off (I can measure 0V) and if I run: `sudo uhubctl -l 2-2 -a on -p 4`
The port power is turned back on (I can measure 5V).
Based on this, I would say, somehow should be possible to switch the port power off, or? Can you please help me?

**Comment by mvp** [comment, marked as answer]
Unfortunately, some USB hubs report per port power switching, but don't really implement voltage (VBUS) control - perhaps to save few cents on circuitry.

This is hardware issue which software cannot possibly fix.

This is main reason why I maintain list of supported devices - every device in that table should actually implement VBUS control.
